### ISSUE TITLE
lint traits with only one implementor

### ISSUE THREAD

**Issue by llogiq** [Manishearth/rust-clippy#1636]
Traits are awesome, as they allow us to decouple the behavior from the data.

However, sometimes they may be overused, leading to indirection and making it harder to read the code than strictly necessary (because one has to look up the trait impls; and there is always the risk other code may also implement those traits).

So, I'd like to (allow for now) lint traits that have only one implementing type, to allow people to quickly identify extension points that have not yet been used, so they can get rid of them and make life easier for rustc and fellow coders.

The lint could either be split in two (for public/private traits), where the former would be `allow`ed, or we allow for aboolean config flag.

**Comment by killercup** [comment]
Interesting idea. Make sure to weight generic type impls differently, though – I want to keep traits like diesel's [FindDsl](http://docs.diesel.rs/diesel/prelude/trait.FindDsl.html) :)

**Comment by shssoichiro** [comment]
I think this makes the most sense to apply to traits that are never exported outside of the crate (i.e. the trait is private or one of its parent modules is private).

**Comment by llogiq** [comment]
Exactly. It may still make sense to habe an `allow`ed Version for public traits – that can help identify extension points that may vor may not be necessary. One could then check out & search all dependent crates for the trait name to identify unused extension points.
